text/0000-interrupt-calling-conventions.md

```diff
@@ -0,0 +1,181 @@
+
+The other issue of the above approach is its performance overhead. Interrupt handlers are often invoked with a very high frequency and at a high priority, so they should be as efficient as possible. However, custom assembly code cannot be optimized by Rust or LLVM, so no inlining or copy elision happens. Also, the wrapper function needs to save all registers that the Rust function could _possibly_ use, because it does not know which registers are actually written by the function.
+
+To avoid these issues, this RFC proposes to add support for _interrupt calling conventions_ to the Rust language. This makes it possible to define interrupt handlers directly as Rust functions, without requiring any wrapper functions or custom assembly code.
+
+Rust already supports three different interrupt calling conventions as experimental features: `msp430-interrupt`, `x86-interrupt`, and `avr-interrupt`. They are already widely used in embedded and operating system kernel projects, so this feature also seems to be useful in practice.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In addition to ABIs for interfacing with external code, Rust also supports so-called _interrupt ABIs_ to define interrupt handler functions that can interface directly with the hardware. These ABIs are only needed for bare-metal applications such as embedded systems or operating system kernels. The ABIs are special because they impose requirements on the whole signature of the function, including arguments and return values.
+
+The following interrupt ABIs are currently supported:
+
+- _(unstable)_ `extern "msp430-interrupt"`: Allows to create interrupt handlers on MSP430 microcontrollers. Functions must have the signature `unsafe extern "msp430-interrupt" fn()`. To add a function to the interrupt table, use the following snippet:
+
+  ```rust
+  #[no_mangle]
+  #[link_section = "__interrupt_vector_10"]
+  pub static TIM0_VECTOR: unsafe extern "msp430-interrupt" fn() = tim0;
+
+  unsafe extern "msp430-interrupt" fn tim0() {...}
+  ```
+
+  Then place the `__interrupt_vector_10` section in the interrupt handler table using a linker script.
+- _(unstable)_ `extern "x86-interrupt"`: This calling convention can be used for defining interrupt handlers on 32-bit and 64-bit `x86` targets. Functions must have one of the following two signatures, depending on the interrupt vector:
+
+  ```rust
+  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame);
+  extern "x86-interrupt" fn(stack_frame: &ExceptionStackFrame, error_code: u64);
```

Comment:
Is the error code also 64bit on 32bit systems?

Reply:
Good catch! I pushed a [new commit](https://github.com/rust-lang/rfcs/pull/3246/commits/719b02c30361106d0523a2b15c816a618103fd94) with some more details on this.